### PR TITLE
Multiple tooltip triggers don't play well together

### PR DESCRIPTION
**Pull request by redbmk**
Fixes issue #16008

### CODE REVIEW

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: a682e7bb8e049679462b67141906b19e5ee323f7
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53424233

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
This breaks on IE8 apparently; see the screencast at https://saucelabs.com/jobs/7ef2d6ebe76540e3a0d10c0053a06a45

File: js/tests/unit/tooltip.js [hunk 1]
```diff
@@ -1180,4 +1180,39 @@ $(function () {
...
+        ['focusin', 'mouseenter', 'focusout', 'mouseleave'],
+
+        ['click', 'focusin', 'mouseenter', 'focusout', 'mouseleave', 'click'],
+        ['mouseenter', 'click', 'focusin', 'focusout', 'mouseleave', 'click'],
+        ['mouseenter', 'focusin', 'click', 'click', 'mouseleave', 'focusout']
+    ]
+
+    assert.ok(!showingTooltip())
+
+    tests.forEach(function (triggers) {
```
Comment (cvrebert) [inline js/tests/unit/tooltip.js]: `Array.forEach` is only available in IE9+; you'll need to use a different approach.

File: js/tests/unit/tooltip.js [hunk 2]
```diff
@@ -1180,4 +1180,39 @@ $(function () {
     assert.strictEqual($tooltip.data('bs.tooltip'), undefined, 'should not initialize the tooltip')
   })
 
+  QUnit.test('should not remove tooltip if multiple triggers are set and one is still active', function (assert) {
+    assert.expect(41)
+    var $el = $('<button/>').bootstrapTooltip({ trigger: 'click hover focus', animation: false })
```
Comment (cvrebert) [inline js/tests/unit/tooltip.js]: * Use proper HTML: `$('<button>Trigger</button>')`
* Add `.appendTo('#qunit-fixture')` like in the other tests

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 5d196bc88fa77868de725f3e51c8790c4cca0db9
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53436671

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by redbmk** [comment]
@cvrebert, I've always just assumed that jQuery handled converting something like "<button/>" to proper elements...I never realized that would break on older browsers. I just pushed your recommended changes.

Assuming this makes sauce labs pass, what's the best way to go forward? Should I create a new pull request with these two commits squashed?

**Comment by redbmk** [comment]
Ugh, that didn't fix it. Looks like IE8 doesn't support forEach so it's probably that.

File: js/tests/unit/tooltip.js (hunk 1 above)
Comment (redbmk) [inline js/tests/unit/tooltip.js]: Oops. Didn't see this comment earlier.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 14e05c2cb1779725ba911283e24a209d6ffd3044
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53437039

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by redbmk** [comment]
OK, looks like everything is good now. Apparently if I squash the commits and do a force-push github will recognize the changes and the pull request will change to a single commit. Is that best practice or is there a better way to do that?

**Comment by cvrebert** [comment]
@redbmk That's our preferred practice. Although we're happy to do the rebasing for you if you don't feel comfortable doing it yourself.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: b31009319e515585198b4462fa07a97de8b32b5c
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53468770

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by redbmk** [comment]
@cvrebert, No worries. I squashed the commits and pulled in the lastest changes from master.

**Comment by mdo** [comment]
Multiple triggers, as in activating a tooltip from two or more methods on the same element?

**Comment by redbmk** [comment]
@mdo Yes - for example the default is "focus hover", but currently if you are both focused and hovered, then as soon as you blur or move the mouse out the tooltip goes away.

With this change, the tooltip will remain as long as you are still either focused or hovered. Clicking (if you have "click" as one of your triggers) will also toggle, so if you click once it will stay showing until you click again, even if you are also using other triggers.

File: js/tooltip.js [hunk 3]
```diff
@@ -109,16 +110,20 @@
...
     if (!self) {
       self = new this.constructor(obj.currentTarget, this.getDelegateOptions())
       $(obj.currentTarget).data('bs.' + this.type, self)
     }
 
+    if (obj instanceof $.Event) {
+      self.inState[obj.type == 'focusin' ? 'focus' : 'hover'] = true
+    }
+
+    if (self.tip().hasClass('in') || (self.hoverState == 'in')) {
```
Comment (fat) [inline js/tooltip.js]: don't need params around `(self.hoverState == 'in')`

File: js/tooltip.js [hunk 4]
```diff
@@ -51,6 +51,7 @@
     this.$element  = $(element)
     this.options   = this.getOptions(options)
     this.$viewport = this.options.viewport && $(this.options.viewport.selector || this.options.viewport)
+    this.inState   = { click: false, hover: false, focus: false }
```
Comment (fat) [inline js/tooltip.js]: `this.inState = null` should be defined in the constructor

File: js/tooltip.js [hunk 5]
```diff
@@ -139,6 +144,12 @@
       $(obj.currentTarget).data('bs.' + this.type, self)
     }
 
+    if (obj instanceof $.Event) {
+      self.inState[obj.type == 'focusout' ? 'focus' : 'hover'] = false
+    }
+
+    if (self.inState.click || self.inState.focus || self.inState.hover) return
```
Comment (fat) [inline js/tooltip.js]: might scale a bit better to do something like this, but not a huge deal:

```js
for (var key in self.instate) {
  if (self.instate[key]) return
}
```

Comment (fat) [inline js/tooltip.js]: or better, create a method like: "self.isInStateTrue()" which returns a boolean since you use that below as well

**Comment by fat** [comment]
thanks for writing a test for this <3

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 4b269037cb42cce7898d57890db40ef0cc3bd5f8
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/60329627

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by redbmk** [comment]
@fat I made the changes you suggested. I'm seeing some errors with travis-ci but don't really understand them. Are those known issues or something I introduced? Looks like the JS tests all pass still.

**Comment by fat** [comment]
hm looks like th pr builder is failing… maybe you need to rebase? @cvrebert would probably know better

**Comment by cvrebert** [comment]
Nah, looks like npm flakiness. Forced a Travis retry, I expect it will succeed.

**Comment by redbmk** [comment]
@fat @cvrebert Looks like the force-retry fixed it. Thanks

**Comment by cvrebert** [comment]
@fat LGTY?

**Comment by fat** [comment]
seems safe to me

**Pull request merged**